### ISSUE TITLE
Typescript type-safe usage with a wrapper function

### ISSUE THREAD

**Issue by CoreyCole** [awslabs/dynamodb-data-mapper-js#59]
I'm trying to implement a wrapper nestJS service around the `DataMapper`. But, I'm not an expert with generics in Typescript and also unsure about `AsyncIterator`. This is what I have so far:

#### a `DynamoDBDataMapperService`:
```typescript
import { Injectable } from '@nestjs/common';
import { DynamoDB } from 'aws-sdk';
import { DataMapper, ScanParameters } from '@aws/dynamodb-data-mapper';

@Injectable()
export class DynamoDBDataMapperService {
    private client: DynamoDB;
    private mapper: DataMapper;

    constructor() {
        this.client = new DynamoDB({
            region: 'us-west-2',
            endpoint: 'http://localhost:8000',
        });
        this.mapper = new DataMapper({ client: this.client });
    }

    public put<T>(item: T): Promise<T> {
        return this.mapper.put(item);
    }

    /**
     * Scans for an the passed item type
     * @param item generic dynamoDB data mapper annotation type
     * @returns AsyncIterator<T> Intended to be consumed with a `for await ... of` loop.
     */
    public get<T>(item: T): AsyncIterator<T> {
        const scanParameter: ScanParameters = { valueConstructor: item };
        return this.mapper.scan(scanParameter);
    }
}
```

#### and then a nestJS service that uses the wrapper:
```typescript
import { Injectable } from '@nestjs/common';
import { CreateCatDto } from './dto/create-cat.dto';
import { Cat } from './schemas/cat.schema';
import { DynamoDBDataMapperService } from '../dynamodb-data-mapper/dynamodb-data-mapper.service';

@Injectable()
export class CatsService {
  constructor(private readonly mapper: DynamoDBDataMapperService) {}

  async create(createCatDto: CreateCatDto): Promise<Cat> {
    const createdCat = new Cat();
    createdCat.name = createCatDto.name;
    createdCat.age = createCatDto.age;
    createdCat.breed = createCatDto.breed;
    return this.mapper.put(createdCat);
  }

  async findAll(): Promise<Cat[]> {
    return await this.mapper.get(Cat);
  }
}
```

#### I'm running into 2 problems:
  1. When I try to set the return type of `DynamoDBDataMapperService.get()` to `AsyncIterable<T>` I get the error
```
[ts] Cannot find name 'AsyncIterator'.
```
  2. I could just be a noob at generics in Typescript, but I'm not sure how to pass the generic type along to `DataMapper.scan()`
```
[ts]
Type '{ valueConstructor: T; }' is not assignable to type 'ScanParameters<StringToAnyObjectMap>'.
  Type '{ valueConstructor: T; }' is not assignable to type 'CtorBearer<StringToAnyObjectMap>'.
    Types of property 'valueConstructor' are incompatible.
      Type 'T' is not assignable to type 'ZeroArgumentsConstructor<StringToAnyObjectMap>'.
```

#### reproduction case
https://github.com/CoreyCole/nest/tree/sample/dynamodb-data-mapper-annotations/sample/20-dynamodb-data-mapper-annotations

#### questions
I could simply solve my problem by instead exposing the mapper as a public property and then reference it directly in my CatsService, but I'd like to learn how to make this wrapper in a type-safe way. I'd like to submit this sample to nestJS for others trying to use `dynamodb-data-mapper` in a nestJS project using best practices as advised from you guys 😃(why I'm not asking on stack overflow)
1. How do I properly use AsyncIterator<T>? It looks like it's been supported in Typescript since 2.3 and my example project is on 2.8
2. How can I pass the generic type along in a `ScanParameters` object to `DataMapper.scan()`?
3. Do you have any other suggestions to how I might improve my github example to better follow your suggested best-practices?

**Comment by jeskew** [comment]
Hi @CoreyCole,

1. You'll need to add `esnext.asynciterable` to the `lib` section of your project's tsconfig file. Support has been included since TypeScript 2.3, but it's still opt-in (see https://www.typescriptlang.org/docs/handbook/compiler-options.html for more on this point).
2. `scan` will infer a type by default, but you can pass it in (e.g., `mapper.scan<T>(constructor)`). Keep in mind that for `query` and `scan`, you wouldn't pass in an item but instead the constructor for the mapped class that should be used for query and scan results.
3. I'm not really familiar with NestJS or its conventions, but I wouldn't use `get` as the name of a method that will perform a full table scan. (`get` normally indicates that a method with look up a single record.)

To gather all results for a query or scan into an array, you would need to iterate over it. The async iterator returned is lazy -- it will only send requests to DynamoDB as you iterate. To collect them all, you would want to do something like:

```typescript
const results: Array<T> = [];
for await (const item of mapper.scan(MyClass)) {
    results.push(item);
}
```

**Comment by CoreyCole** [comment]
Thanks!
